File: card.cpp
///////////////////////////////////////////////////////////////////////////////
// File:   card.cpp                    Summer 2013
// Author: Patrick Vargas              patrick.vargas@colorado.edu
// Description:
//    Provides a basic card class for use in a card game program or simulation.
///////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////////////////////////////
#include "bicycle.h"
///////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////////////////////////////
namespace bicycle {
  //
  // Constructors
  //
  // Legal Values Include
  //   rank: 1-14; 1 default
  //   suit: SPADES, DIAMONDS, CLUBS, HEARTS; Spades by default
  //   value: Any value based on game or simulation; 0 by default
  //   isFaceUp: Boolean; False (or down) by default
  Card::Card ( short rankInit, short suitInit, bool isFaceUpInit, short valueInit, Point p) {
    rank_     = rankInit;
    suit_     = suitInit;
    value_    = valueInit;
    isFaceUp_ = isFaceUpInit;
    position_ = p;
  }

  //
  // Member Methods // Getters
  // 
  const bool   Card::isFaceUp( ) { return isFaceUp_; }
  const short  Card::rank( )     { return rank_;     }
  const short  Card::suit( )     { return suit_;     }
  const short  Card::value( )    { return value_;    }
  const Point  Card::position( ) { return position_; }

  //
  // Member Methods // Setters
  //
  // Pre: Must be a boolean
  void Card::isFaceUp( bool temp ) {
    assert( temp == true || temp == false );
    isFaceUp_ = temp;
  }

  // Pre: Must Be a value between 1 and 14
  void Card::rank( short temp ) {
    assert( temp >= 1 && temp <= 14 );
    rank_ = temp;
  }

  // Pre: Must be a value in enum Suit
  void Card::suit( short temp ) {
    assert( temp == SPADES || temp == DIAMONDS || temp == CLUBS || temp == HEARTS || temp == JOKERS );
    suit_ = temp;
  }

  // Pre: Can be any value
  void Card::value( short temp ) { value_ = temp; }

  // Pre: Can be any position in the window
  void Card::position( Point temp ) { position_ = temp; }

  //
  // Auxiliary Methods
  //
  // Post: Flips card over
  void Card::flip( ) { isFaceUp ( !isFaceUp_ ); }

  // Post: Prints the necessary information
  void Card::print( ) {
    stringstream ss;
    string rank, suit;
    
    switch ( rank_ ) {
      case 14:  rank = "Joker"; break;
      case 13:  rank = "King";  break;
      case 12:  rank = "Queen"; break;
      case 11:  rank = "Jack";  break;
      case  1:  rank = "Ace";   break;
      default:
        ss << rank_;
        rank = ss.str();
        break;
    }
    
    switch ( suit_ )
    {
      case SPADES:    suit = "Spades";    break;
      case DIAMONDS:  suit = "Diamonds";  break;
      case CLUBS:     suit = "Clubs";     break;
      default:        suit = "Hearts";    break;
    }
    
    cout << rank << " of " << suit;
  }

  // Post: Card has been outputted to ostream
  ostream &operator<<( ostream &output, const Card &c )
  { 
    switch ( c.rank_ ) {
      case 14:  output << "O";     break;
      case 13:  output << "K";     break;
      case 12:  output << "Q";     break;
      case 11:  output << "J";     break;
      case  1:  output << "A";     break;
      default:  output << c.rank_; break;
    }

    // output << " of ";
    
    switch ( c.suit_ )
    {
      case SPADES:    output << "S"; break;
      case DIAMONDS:  output << "D"; break;
      case CLUBS:     output << "C"; break;
      default:        output << "H"; break;
    }

    return output;            
  }
}
///////////////////////////////////////////////////////////////////////////////

File: src/Card.java
/**
 * Created by pevargas90 on 4/18/14.
 *
 * Provides a basic card class for use in a card game program or simulation.
 */
public class Card {
    private Integer rank_;
    private Suit    suit_;
    private Integer value_;
    private Boolean is_face_up_;

    /**
     * Constructor
     */
    public Card () {
        rank_       = 1;
        suit_       = Suit.SPADES;
        value_      = 0;
        is_face_up_ = false;
    }

    /**
     * Constructor
     *
     * @param rank The Initial Rank
     * @param suit The Initial Suit
     */
    public Card (int rank, Suit suit) {
        rank_       = rank;
        suit_       = suit;
        value_      = 0;
        is_face_up_ = false;
    }

    /**
     * Constructor
     *
     * @param rank The Initial Rank
     * @param suit The Initial Suit
     * @param value The Initial Value
     */
    public Card (int rank, Suit suit, int value) {
        rank_       = rank;
        suit_       = suit;
        value_      = value;
        is_face_up_ = false;
    }

    /**
     * Gets the rank
     * @return The rank has been returned.
     */
    public int rank() {
        return rank_;
    }

    /**
     * Sets the rank
     * @param rank The rank value
     */
    public void rank(int rank) {
        rank_ = rank;
    }

    /**
     * Gets the suit
     * @return The suit has been returned
     * @see Suit
     */
    public Suit suit() {
        return suit_;
    }

    /**
     * Sets the suit
     * @param suit The suit value
     * @see Suit
     */
    public void suit(Suit suit) {
        suit_ = suit;
    }

    /**
     * Gets the value
     * @return The value has been returned
     */
    public int value() {
        return value_;
    }

    /**
     * Sets the value
     * @param value The value value
     */
    public void value(int value) {
        value_ = value;
    }

    /**
     * Gets the is face up boolean
     * @return Returns if the card is face up or not
     */
    public boolean is_face_up() {
        return is_face_up_;
    }

    /**
     * Sets the face up value
     * @param is_face_up True or False
     */
    public void is_face_up(boolean is_face_up) {
        is_face_up_ = is_face_up;
    }

    /**
     * Flips the card over
     */
    public void flip() {
        is_face_up_ = !is_face_up_;
    }

    /**
     * Converts card to string representation
     * @return String of information
     */
    public String toString() {
        String temp;

        switch ( rank_ ) {
            case 14: temp = "O"; break;
            case 13: temp = "K"; break;
            case 12: temp = "Q"; break;
            case 11: temp = "J"; break;
            case 1:  temp = "A"; break;
            default: temp = rank_.toString(); break;
        }
        switch ( suit_ ) {
            case SPADES:   temp += "S"; break;
            case DIAMONDS: temp += "D"; break;
            case CLUBS:    temp += "C"; break;
            case HEARTS:   temp += "H"; break;
            case JOKERS:   temp += "J"; break;
            default:       temp += "O"; break;
        }

        return temp;
    }

    /**
     * Prints to command line
     */
    public void print() {
        String temp;
        switch ( rank_ ) {
            case 14: temp = "Joker"; break;
            case 13: temp = "King"; break;
            case 12: temp = "Queen"; break;
            case 11: temp = "Jack"; break;
            case 1:  temp = "Ace"; break;
            default: temp = rank_.toString(); break;
        }
        System.out.print(temp + " of ");

        switch ( suit_ ) {
            case SPADES:   temp = "Spades"; break;
            case DIAMONDS: temp = "Diamonds"; break;
            case CLUBS:    temp = "Clubs"; break;
            case HEARTS:   temp = "Hearts"; break;
            case JOKERS:   temp = "Jokers"; break;
            default:       temp = "Other"; break;
        }
        System.out.println(temp);
    }
}
